rir/src/compiler/translations/pir_2_rir/pir_2_rir.cpp

```diff
@@ -733,7 +736,8 @@ rir::Code* Pir2Rir::compileCode(Context& ctx, Code* code) {
                 auto mkenv = MkEnv::Cast(stvar->env());
                 assert(!MkEnv::Cast(stvar->arg(0).val()));
                 if (stvar->isStArg) {
-                    if (mkenv && mkenv->stub) {
+                    if (mkenv && mkenv->stub &&
+                        (!stvar->bb()->isDeopt() || !seenMkArg)) {
```

Comment:
I do not see how a change in the continue branch could happen. My change applies only on how `stVars` are emitted in deopt branches.

Reply:
be creative...

Assuming BB's are visited: S,A,B,C,D
```
S:
  e = mkenv
  checkpoint -> A B
A:
  mkarg(..., e)
  ...
B:
  checkpoint -> C D
C:
  stvar(..., e)  # this will emit stvar, when it should emit stvar stubbed  
  ...
D:
  ...
```

Reply:
I must be missing sth, but from what I see, the `LoweringVisitor` at some point will `visit C (after S,A,B)` and since `mkenv && mkenv->stub && !stvar->bb()->isDeopt()` it would emit a `stVarStubbed`.

Reply:
>  it would emit a stVarStubbed.

no it won't since seenMkarg is true

Reply:
details aside, you can't use a global variable to track a flow-sensitive property...

Reply:
But...
1. It is an `||` clause. 
2. I am not using a global. `seenMkArg` is a per BB var, it should be restarted on every `bb` as far as I understand the visitor. And the only idea is, if I am in a deopt block, in which an `mkArg` already happened, I should not emit store/load stubs, because the mkArg would have already materialized the env.

Reply:
>  seenMkArg is a per BB var, 

oh, I missed that.

well then it will break for mkarg in the fastpath (i know, currently not, but ugh, this is brittle) and deopt branches with more than one BB. All our deopts use at least 2 BBs after lower() in the pir2rir.

Reply:
ignoring if this particular implementation might work with the kind of code we generate, still:

you are racking a flow sensitive property, with a condition that is not flow sensitive.

don't you agree that there is a design flaw here?

Reply:
Definitely, my assumption was that `deopt` branches consist always on a single `basic block`. If that is not the case, then this will not work

Reply:
Oth, there are lot of places on the code that uses the `isDeoptBanch` which only looks for a `deopt` instruction in the branch. At least, that is how I've been using it. This opt, only elides an env in case the `mkArg` is inside a single `deoptBranch`, and then the change when lowering `ldVars and stVars` also only applies to that single `block`. 
The opt could be explain as: 
`We can elide environments even if there are `mkArgs` using it but those must only reside inside basic blocks but only if they are deopt branches (contain a deopt isntruction). In that case, in those deopt branches we must ensure that ldVars and stVars after the mkArg instruction (which now materializes de env silently) use the real materialized environment instead of the stub`

Reply:
currently the following is the case:

before Pir2Rir::lower  :  deopt branches are (probably) always 1BB
after Pir2Rir::lower : deopt branches are always at least 2 BBs (because assume gets converted to branch).

Reply:
it would be good if we don't rely too much on how exactly the deopt branch looks. I know about the issue with isDeoptBranch. we should fix that method to better deal with mulit-bb deopt branches.

Reply:
I don't want to include a transformation that relies on the fact that we do not split a BB into two. that is beyond silly.

Reply:
what would be the problem with adding a materialize env instruction at the pir level? sth. that takes a sutb env as input and returns an env as output. that's probably easier to implement than the time spent thinking about in which situations the current implementation might be correct...